File: samples/bot-tag-mention/nodejs/dialogs/mainDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ConfirmPrompt, DialogSet, DialogTurnStatus, OAuthPrompt, WaterfallDialog} = require('botbuilder-dialogs');
const { LogoutDialog } = require('./logoutDialog');

const CONFIRM_PROMPT = 'ConfirmPrompt';
const MAIN_DIALOG = 'MainDialog';
const MAIN_WATERFALL_DIALOG = 'MainWaterfallDialog';
const OAUTH_PROMPT = 'OAuthPrompt';
const { SimpleGraphClient } = require('../simpleGraphClient');
const { polyfills } = require('isomorphic-fetch');
const { CardFactory, TurnContext} = require('botbuilder-core');
const { TeamsInfo } = require('botbuilder');

class MainDialog extends LogoutDialog {
    constructor() {
        super(MAIN_DIALOG, process.env.connectionName);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, {
            connectionName: process.env.connectionName,
            text: 'Please Sign In',
            title: 'Sign In',
            timeout: 300000
        }));
        this.addDialog(new ConfirmPrompt(CONFIRM_PROMPT));
        this.addDialog(new WaterfallDialog(MAIN_WATERFALL_DIALOG, [
            this.promptStep.bind(this),
            this.mentionTag.bind(this),
            this.ensureOAuth.bind(this)
        ]));

        this.initialDialogId = MAIN_WATERFALL_DIALOG;
    }

    /**
     * The run method handles the incoming activity (in the form of a DialogContext) and passes it through the dialog system.
     * If no dialog is active, it will start the default dialog.
     * @param {*} dialogContext
     */
    async run(context, accessor) {
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this);
        const dialogContext = await dialogSet.createContext(context);
        const results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dialogContext.beginDialog(this.id);
        }
    }

    async promptStep(stepContext) {
        try {
            return await stepContext.beginDialog(OAUTH_PROMPT);
        } catch (err) {
            console.error(err);
        }
    }

    // Sends tag mention adaptive card.
    async tagMentionAdaptiveCard(stepContext, tagName, tagId) {
        const adaptiveCardAttachment = {
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "type": "AdaptiveCard",
            "version": "1.5",
            "speak": "This card mentions a tag:  "+tagName,
            "body": [
              {
                "type": "TextBlock",
                "text": "Mention a Tag: <at>Test Tag</at>"
              }
            ],
            "msteams": {
              "entities": [
                {
                  "type": "mention",
                  "text": "<at>Test Tag</at>",
                  "mentioned": {
                    "id": tagId,
                    "name": tagName,
                    "type": "tag"
                  }
                }
              ]
            }
          }
        await stepContext.context.sendActivity({
            attachments: [CardFactory.adaptiveCard(adaptiveCardAttachment)]
        });
    }

    // Method to invoke Tag mention functionality flow.
    async mentionTag(stepContext) {
        try {
            const tokenResponse = stepContext.result;
            if (stepContext.context.activity.conversation.conversationType === 'personal' && tokenResponse.token) {
                await stepContext.context.sendActivity('You have successfully logged in. Please install the app in the team scope to use the Tag mention functionality.');
            } else {
                let tagExists = false;
                if (tokenResponse.token) {
                    TurnContext.removeRecipientMention(stepContext.context._activity);
                    const activityText = stepContext.context.activity.text.trim();
                    if (activityText.includes('<at>')) {
                        const tagName = activityText.replace('<at>', '').replace('</at>', '').trim();
                        const tagID = stepContext.context._activity.entities[1].mentioned.id;
                        await this.tagMentionAdaptiveCard(stepContext, tagName, tagID);
                    } else if (activityText !== '') {
                        let client, teamDetails;
                        try {
                            // Fetch data from Microsoft Graph.
                            client = new SimpleGraphClient(tokenResponse.token);
                            teamDetails = await TeamsInfo.getTeamDetails(stepContext.context);
                        } catch (ex) {
                            await stepContext.context.sendActivity('You don\'t have Graph API permissions to fetch tag\'s information. Please use this command to mention a tag: "`@<Bot-name>  @<your-tag>`" to experience tag mention using the bot.');
                        }
                        const result = await client.getTag(teamDetails.aadGroupId);
                        for (const tagDetails of result.value) {
                            if (tagDetails.displayName === activityText) {
                                tagExists = true;
                                await this.tagMentionAdaptiveCard(stepContext, tagDetails.displayName, tagDetails.id);
                                break;
                            }
                        }
                        if (!tagExists) {
                            await stepContext.context.sendActivity('Provided tag name is not available in this team. Please try with another tag name or create a new tag.');
                        }
                    } else {
                        await stepContext.context.sendActivity('Please provide a tag name while mentioning the bot as "`@<Bot-name> <your-tag-name>`" or mention a tag as "`@<Bot-name> @<your-tag>`"');
                    }
                } else {
                    console.log('Response token is null or empty.');
                }
            }
        } catch (ex) {
            console.error('Error occurred while processing your request.', ex.message);
        }
    
        return await stepContext.endDialog();
    }

    async ensureOAuth(stepContext) {
        await stepContext.context.sendActivity('Thank you.');

        const result = stepContext.result;
        if (result) {
            // Call the prompt again because we need the token. The reasons for this are:
            // 1. If the user is already logged in we do not need to store the token locally in the bot and worry
            // about refreshing it. We can always just call the prompt again to get the token.
            // 2. We never know how long it will take a user to respond. By the time the
            // user responds the token may have expired. The user would then be prompted to login again.
            //
            // There is no reason to store the token locally in the bot because we can always just call
            // the OAuth prompt to get the token or get a new token if needed.
            return await stepContext.beginDialog(OAUTH_PROMPT);
        }
        return await stepContext.endDialog();
    }

}

module.exports.MainDialog = MainDialog;

File: samples/bot-tag-mention/python/dialogs/main_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import MessageFactory
from botbuilder.dialogs import (
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
    PromptOptions,
)
from botbuilder.dialogs.prompts import OAuthPrompt, OAuthPromptSettings, ConfirmPrompt

from .logout_dialog import LogoutDialog
from botbuilder.core import CardFactory, TurnContext
from botbuilder.core.teams.teams_info import TeamsInfo
from simple_graph_client import SimpleGraphClient
import logging

logging.basicConfig(level=logging.INFO)

class MainDialog(LogoutDialog):
    def __init__(self, connection_name: str):
        # Initializes the MainDialog with OAuthPrompt, ConfirmPrompt, and WaterFallDialog.
        super(MainDialog, self).__init__(MainDialog.__name__, connection_name)

        self.add_dialog(
            OAuthPrompt(
                OAuthPrompt.__name__,
                OAuthPromptSettings(
                    connection_name=connection_name,
                    title="Please Sign In",
                    timeout=300000
                )
            )
        )

        self.add_dialog(ConfirmPrompt(ConfirmPrompt.__name__))

        self.add_dialog(
            WaterfallDialog(
                "WFDialog",
                [
                    self.prompt_step,
                    self.login_step,
                    self.mention_tag,
                ],
            )
        )

        self.initial_dialog_id = "WFDialog"

    # Prompts the user to sign in using the OAuthPrompt.
    async def prompt_step(self, step_context):
        logging.info("Reached OAuthPrompt step - prompting for login.")
        return await step_context.begin_dialog(OAuthPrompt.__name__)
    
    # Handles the login result and moves to tag mention step.
    async def login_step(self, step_context):
        if step_context.result and step_context.result.token:
            logging.info("Login Step Reached — Successfully Logged In")
            if step_context.context.activity.conversation.conversation_type == 'personal':
                await step_context.context.send_activity(
                    "You are logged in. Please install the app in the team scope to use the Tag mention functionality."
                )
            return await step_context.next(step_context.result)  # Pass token to next step

        # Handles invalid actions or cancelled login
        await step_context.context.send_activity(
            "Login was not successful or was cancelled. Let's try again."
        )
        return await step_context.replace_dialog(self.initial_dialog_id)

    # Sends tag mention adaptive card.
    async def tagMentionAdaptiveCard(self, step_context: WaterfallStepContext, tagName: str, tagId: str) -> DialogTurnResult:
        adaptiveCardAttachment = {
            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
            "type": "AdaptiveCard",
            "version": "1.5",
            "speak": f"This card mentions a tag: {tagName}",
            "body": [
                {
                    "type": "TextBlock",
                    "text": f"Mention a Tag: <at>{tagName}</at>"
                }
            ],
            "msteams": {
                "entities": [
                    {
                        "type": "mention",
                        "text": f"<at>{tagName}</at>",
                        "mentioned": {
                            "id": tagId,
                            "name": tagName,
                            "type": "tag"
                        }
                    }
                ]
            }
        }
        await step_context.context.send_activity(
            MessageFactory.attachment(CardFactory.adaptive_card(adaptiveCardAttachment))
        )
        return await step_context.end_dialog()
    

    # Method to invoke Tag mention functionality flow.
    async def mention_tag(self, step_context):
        try:
            tokenResponse = step_context.result
            if step_context.context.activity.conversation.conversation_type == 'personal' and tokenResponse and tokenResponse.token:
               await step_context.context.send_activity('You have successfully logged in. Please install the app in the team scope to use the Tag mention functionality.') 
            else:
                tagExists = False
                if tokenResponse and tokenResponse.token:
                    TurnContext.remove_recipient_mention(step_context.context.activity)
                    activityText = step_context.context.activity.text.strip()
                    if '<at>' in activityText:
                        tagName = activityText.replace('<at>', '').replace('</at>', '').strip()
                        if len(step_context.context.activity.entities) > 1:
                            tagID = step_context.context.activity.entities[1].additional_properties.get('mentioned', {}).get('id')
                            await self.tagMentionAdaptiveCard(step_context, tagName, tagID)
                        else:
                            await step_context.context.send_activity("Could not find tag information in the message.")
                    elif activityText != '':
                        try:
                            # Fetch data from Microsoft Graph.
                            client = SimpleGraphClient(tokenResponse.token)
                            teamDetails = await TeamsInfo.get_team_details(step_context.context)
                            result = await client.get_Tag(teamDetails.aad_group_id)
                            
                            for tagDetails in result.get('value', []):
                                if tagDetails.get('displayName') == activityText:
                                    tagExists = True
                                    await self.tagMentionAdaptiveCard(step_context, tagDetails.get('displayName'), tagDetails.get('id'))
                                    break
                                    
                            if not tagExists:
                                await step_context.context.send_activity(
                                    "Provided tag name is not available in this team. Please try with another tag name or create a new tag."
                                )
                        except Exception as e:
                            logging.error(f"Error fetching team details: {e}")
                            await step_context.context.send_activity("You don't have Graph API permissions to fetch tag's information. Please use this command to mention a tag: '`@<Bot-name>  @<your-tag>`' to experience tag mention using the bot.")
                    else:
                        await step_context.context.send_activity(
                            'Please provide a tag name while mentioning the bot as "`@<Bot-name> <your-tag-name>`" or mention a tag as "`@<Bot-name> @<your-tag>`"'
                        )
                else:
                    await step_context.context.send_activity(
                        'Response token is null or empty.'
                    )
        except Exception as e:
            logging.error(f'Error occurred while processing your request: {e}')
            await step_context.context.send_activity('An error occurred while processing your request.')

        return await step_context.end_dialog()

